### PR TITLE
Feature preprocessors, Loss strategies

### PR DESCRIPTION
**Pull request by ravinkohli**
Added the following components from old autopytorch code:

1. Feature Preprocessors like Kitchen sinks, FastICA, KernelPCA, RandomKitchenSinks, Nystroem, PolynomialFeatures,         PowerTransformer, TruncatedSVD
2. Weighted Loss strategies for binary and multiclass classification

### CODE REVIEW

File: autoPyTorch/datasets/base_dataset.py [hunk 1]
```diff
@@ -59,17 +60,17 @@ def __getitem__(self, idx: int) -> np.ndarray:
...
-        dataset_name: Optional[str] = None,
-        val_tensors: Optional[BASE_DATASET_INPUT] = None,
-        test_tensors: Optional[BASE_DATASET_INPUT] = None,
-        resampling_strategy: Union[CrossValTypes, HoldoutValTypes] = HoldoutValTypes.holdout_validation,
-        resampling_strategy_args: Optional[Dict[str, Any]] = None,
-        shuffle: Optional[bool] = True,
-        seed: Optional[int] = 42,
-        train_transforms: Optional[torchvision.transforms.Compose] = None,
-        val_transforms: Optional[torchvision.transforms.Compose] = None,
+            self,
```
Comment (franchuterivera) [inline autoPyTorch/datasets/base_dataset.py:63]: I also so this indentation change in Sebastian's code :), does it come from pycharm? I think probably you should change the settings for this...

File: autoPyTorch/datasets/base_dataset.py [hunk 2]
```diff
@@ -245,9 +246,9 @@ def get_splits_from_resampling_strategy(self) -> List[Tuple[List[int], List[int]
         return splits
 
     def create_cross_val_splits(
-        self,
-        cross_val_type: CrossValTypes,
-        num_splits: int
+            self,
```
Comment (franchuterivera) [inline autoPyTorch/datasets/base_dataset.py:249]: same here

File: autoPyTorch/datasets/base_dataset.py [hunk 3]
```diff
@@ -277,9 +278,9 @@ def create_cross_val_splits(
         return splits
 
     def create_holdout_val_split(
-        self,
-        holdout_val_type: HoldoutValTypes,
-        val_share: float,
+            self,
```
Comment (franchuterivera) [inline autoPyTorch/datasets/base_dataset.py:281]: same here

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    base_feature_preprocessor import autoPyTorchFeaturePreprocessingComponent
+
+preprocessing_directory = os.path.split(__file__)[0]
+_preprocessors = find_components(__package__,
+                                 preprocessing_directory,
+                                 autoPyTorchFeaturePreprocessingComponent)
+_addons = ThirdPartyComponents(autoPyTorchFeaturePreprocessingComponent)
+
+
+def add_encoder(encoder: autoPyTorchFeaturePreprocessingComponent) -> None:
```
Comment (franchuterivera) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py:24]: Should this be called encoder?

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    _addons.add_component(encoder)
+
+
+class FeatureProprocessorChoice(autoPyTorchChoice):
+    """
+    Allows for dynamically choosing encoding component at runtime
```
Comment (franchuterivera) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py:30]: probably we should update the doc string

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    """
+
+    def get_components(self) -> Dict[str, autoPyTorchComponent]:
+        """Returns the available encoder components
```
Comment (franchuterivera) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py:34]: here too

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            preprocessor = CSH.CategoricalHyperparameter('__choice__',
+                                                         list(available_.keys()),
+                                                         default_value=default)
+
+        cs.add_hyperparameter(preprocessor)
+
+        # add only child hyperparameters of early_preprocessor choices
+        for name in preprocessor.choices:
+            updates = self._get_search_space_updates(prefix=name)
+            config_space = available_[name].get_hyperparameter_search_space(dataset_properties,  # type:ignore
```
Comment (franchuterivera) [inline autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py:102]: why type ignore here?

File: autoPyTorch/datasets/base_dataset.py (hunk 1 above)
Comment (ravinkohli) [reply autoPyTorch/datasets/base_dataset.py:63]: yeah I think so

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py (hunk 4 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py:24]: thanks for pointing it out

File: autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py (hunk 7 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/feature_preprocessing/base_feature_preprocessor_choice.py:102]: so as get_hyperparameter_search_space of the base component only has dataset properties as a parameter, it gives a "too many arguments for get_hyperparameter search space" error. If I add a **kwargs to the base component, it gives signature invalid for every class inheriting from base component. On searching on google, I found [this](https://github.com/python/mypy/issues/5364)

**Review by franchuterivera — APPROVED** [review]

**Pull request merged**